include/RAJA/policy/sycl/kernel/internal.hpp

```diff
@@ -95,8 +95,53 @@ struct LaunchDims {
     launch_local.y = std::max(launch_local.y, local.y);
     launch_local.z = std::max(launch_local.z, local.z);
 
+    cl::sycl::queue* q = ::RAJA::sycl::detail::getQueue();
+    // Global resource was not set, use the resource that was passed to forall
+    // Determine if the default SYCL res is being used
+    if (!q) {
+      camp::resources::Resource sycl_res = camp::resources::Sycl();
+      q = sycl_res.get<camp::resources::Sycl>().get_queue();
+    }
+
+    cl::sycl::device dev = q->get_device();
+
+    auto max_work_group_size = dev.get_info< ::cl::sycl::info::device::max_work_group_size>();
+
+    if(launch_local.x > max_work_group_size) {
+      launch_local.x = max_work_group_size;
+    }
+    if(launch_local.y > max_work_group_size) {
+      launch_local.y = max_work_group_size;
+    }
+    if(launch_local.z > max_work_group_size) {
+      launch_local.z = max_work_group_size;
+    }
 
-// User gave group policy, use to calculate global space
+
+    // Make sure the multiple of locals fits
+    // Prefer larger z -> y -> x
+    if(launch_local.x * launch_local.y * launch_local.z > max_work_group_size) {
+      int remaining = 1;
+      // local z cannot be > max_wrk from above
+      // if equal then remaining is 1, on handle < 
+      if(max_work_group_size > launch_local.z) {
+        // keep local z
+        remaining = max_work_group_size / launch_local.z;
+      }
+      if(remaining >= launch_local.y) {
+        // keep local y
+        remaining = remaining / launch_local.y;
+      } else {
+        launch_local.y = remaining;
+        remaining = remaining / launch_local.y;
```

Comment:
Won't this cause remaining to be 0?

Reply:
When we are already at the max_work_group_size it will be 1.

Reply:
In this else `remaining < launch_local.y` so won't `remaining / launch_local.y` always be 0?

Reply:
`launch_local.y` is being assigned with the remaining value first.  line 136 could be changed to `remaining = 1;` as it ends up being `remaining / remaining`.  I wrote it this way to be more in line with the surrounding logic, but I can make the change if it makes things clearer.

Reply:
Oh, my bad, thanks for showing me that.